### PR TITLE
feat: [CDS-41589]: variables form for deploymentTemplate + customVariableEditable to support description 

### PR DESCRIPTION
**Pull request by PratyushGar**
…ableEditable to support description + template label corrected on list view

#### Screenshots

image

image

image

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 59 Issues 
* Code issue 59 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/10842)

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx [hunk 1]
```diff
@@ -189,6 +189,7 @@ export class CustomVariables extends Step<CustomVariablesData> {
         ...(!isNil(row.default)
           ? { default: row.type === 'Number' ? parseFloat(row.default as unknown as string) : row.default }
           : {}),
+        ...(!!isDescriptionEnabled && { description: row.description }),
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx:192]: just check if `!isNil(row.description)`

File: src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/Cards/DeploymentTemplateCard.tsx [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import { VariableType } from '@pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableUtils'
+import VariableAccordionSummary from '../VariableAccordionSummary'
+import type { PipelineVariablesData } from '../types'
+import css from '../PipelineVariables.module.scss'
+import moduleCss from './DeploymentTemplateCard.module.scss'
+
+export interface DeploymentTemplateCardProps {
+  deploymentTemplate: any
+  originalDeploymentTemplate: any
+  unresolvedDeploymentTemplate: any
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/Cards/DeploymentTemplateCard.tsx:34]: These 3 should be of type `DeploymentConfig`

File: src/modules/72-templates-library/components/TemplateStudio/TemplateVariables/TemplateVariables.tsx [hunk 3]
```diff
@@ -46,7 +47,8 @@ const TemplateVariables: React.FC = (): JSX.Element => {
 
   const onUpdate = useCallback(
     async (values: PipelineInfoConfig | StageElementConfig | StepElementConfig) => {
-      const processNode = omit(values, 'name', 'identifier', 'description', 'tags')
+      const isDeploymentTemplate = originalTemplate.type === TemplateType.CustomDeployment
+      const processNode = omit(values, 'name', 'identifier', !isDeploymentTemplate ? 'description' : '', 'tags')
```
Comment (yc147) [inline src/modules/72-templates-library/components/TemplateStudio/TemplateVariables/TemplateVariables.tsx:51]: Why is this needed?

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/DeploymentInfraWrapper/DeploymentInfraSpecifications/DeploymentInfraSpecifications.tsx [hunk 4]
```diff
@@ -251,7 +122,7 @@ export default function DeploymentInfraSpecifications(props: { formik: FormikPro
         headerClassName={css.headerText}
       >
         <div>
-          <Layout.Horizontal flex={{ alignItems: 'flex-start' }} margin={{ bottom: 'medium' }}>
+          <Layout.Horizontal className={css.fetchInstanceScript}>
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/DeploymentInfraWrapper/DeploymentInfraSpecifications/DeploymentInfraSpecifications.tsx:125]: Using props with styled component is faster then using className

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/DeploymentInfraWrapper/DeploymentInfraWrapper.tsx [hunk 5]
```diff
@@ -69,18 +69,18 @@ export const DeploymentInfraWrapper = ({ children }: React.PropsWithChildren<unk
...
     })
-    updateTemplate(updatedConfig)
+    updateDeploymentConfig(updatedConfig)
   }
 
   // eslint-disable-next-line react-hooks/exhaustive-deps
   const debounceUpdateTemplate = React.useCallback(
     debounce((infraValues: DeploymentInfra) => updateConfigValue(infraValues), 300),
-    [updateTemplate]
+    [updateDeploymentConfig]
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/DeploymentInfraWrapper/DeploymentInfraWrapper.tsx:83]: This should be `updateConfigValue`

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableEditable.tsx [hunk 6]
```diff
@@ -237,14 +253,36 @@ export function CustomVariableEditable(props: CustomVariableEditableProps): Reac
...
+                      )}
                       <div
                         className={cx(css.valueRow, {
                           [css.selectedSearchTextValueRow]: searchText?.length && isValidValueMatch,
                           'selected-search-text': searchedEntityType === 'value' && hasSameMetaPath
                         })}
                       >
                         <div>
-                          {variable.type === VariableType.Secret ? (
+                          {(variable?.type as any) === VariableType.Connector ? (
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableEditable.tsx:270]: Isn't using `unknown` better than `any` in this case?

File: src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx [hunk 7]
```diff
@@ -110,13 +114,10 @@ export function TemplateVariablesContextProvider(
 
   React.useEffect(() => {
     const templateType =
-      resolvedTemplate.type?.toLowerCase() === monitoredServiceYamlKey.toLowerCase()
-        ? monitoredServiceYamlKey
-        : resolvedTemplate.type?.toLowerCase() === secretManagerYamlKey.toLowerCase()
-        ? secretManagerYamlKey
-        : resolvedTemplate.type?.toLowerCase()
+      templateTypeYamlKey[resolvedTemplate.type?.toLowerCase()] || resolvedTemplate.type?.toLowerCase()
```
Comment (yc147) [inline src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx:117]: const templateLabel = templateFactory.getTemplateLabel(template.type).replace(/\s/g,'')
const templateVariablesKey = templateLabel[0].toLowerCase() + templateLabel.slice(1);

Can we do something like this?

**Review by yc147 — COMMENTED** [review]
Let's also fix the css for variables panel

File: src/modules/72-templates-library/components/TemplateStudio/TemplateVariables/TemplateVariables.tsx (hunk 3 above)
Comment (PratyushGar) [reply src/modules/72-templates-library/components/TemplateStudio/TemplateVariables/TemplateVariables.tsx:52]: In existing code, the `description was omitted before sanitizing`, so added a conditional check, because same data is called with `updateTemplate`

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx (hunk 1 above)
Comment (PratyushGar) [reply src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx:192]: processNode parses data and passes to the `onUpdate` function which in some places is not using sanitize to omit `description` while updating the entity, so added conditional check

File: src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx (hunk 7 above)
Comment (PratyushGar) [reply src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx:117]: `Label` can be different for templates, `templateVariablesKeyMap` is camelcase of type so added a map for the same.

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariablesEditableStage.tsx [hunk 8]
```diff
@@ -83,7 +83,7 @@ export function CustomVariablesEditableStage(props: CustomVariableEditableProps)
 
   function addNew(): void {
     setSelectedVariable({
-      variable: { name: '', type: 'String', value: '' },
+      variable: { name: '', type: 'String', value: '', ...(isDescriptionEnabled && { description: '-' }) },
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariablesEditableStage.tsx:86]: Adding default `-` value for description when flag enabled, to prevent in other use cases YAML.

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/DeploymentInfraWrapper/DeploymentInfraWrapper.tsx [hunk 9]
```diff
@@ -60,12 +60,12 @@ export const DeploymentInfraWrapper = ({ children }: React.PropsWithChildren<unk
         set(
           draft,
           'variables',
-          draft?.variables?.filter(variable => variable.value).map(({ id, ...variable }) => variable)
```
Comment (PratyushGar) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/DeploymentInfraWrapper/DeploymentInfraWrapper.tsx:62]: removed filter to enable user to save variable even if variable value is not filled
